src/kvstore/CompactionFilter.h

```diff
@@ -80,18 +80,12 @@ class KVCompactionFilterFactory : public rocksdb::CompactionFilterFactory {
     auto now = time::WallClock::fastNowInSec();
     if (context.is_full_compaction || context.is_manual_compaction) {
       LOG(INFO) << "Do full/manual compaction!";
-      lastRunCustomFilterTimeSec_ = now;
-      return std::make_unique<KVCompactionFilter>(spaceId_, createKVFilter());
     } else {
-      if (FLAGS_custom_filter_interval_secs >= 0 &&
-          now - lastRunCustomFilterTimeSec_ > FLAGS_custom_filter_interval_secs) {
-        LOG(INFO) << "Do custom minor compaction!";
-        lastRunCustomFilterTimeSec_ = now;
-        return std::make_unique<KVCompactionFilter>(spaceId_, createKVFilter());
-      }
-      LOG(INFO) << "Do default minor compaction!";
-      return std::unique_ptr<rocksdb::CompactionFilter>(nullptr);
+      // No worry, by default flush will not go through the custom compaction filter.
+      // See CompactionFilterFactory::ShouldFilterTableFileCreation.
+      LOG(INFO) << "Do automatic or periodic compaction!";
     }
+    return std::make_unique<KVCompactionFilter>(spaceId_, createKVFilter());
```

Comment:
The modification will make every compaction go through `CompactionFilter`, it will hurt performance a lot. As your issue, I suggest you modify the options `periodic_compaction_seconds` in RocksDB. 

See `kBottommostFiles` in RocksDB

Reply:
> The modification will make every compaction go through `CompactionFilter`, it will hurt performance a lot. As your issue, I suggest you modify the options `periodic_compaction_seconds` in RocksDB.
> 
> See `kBottommostFiles` or `kPeriodicCompaction` in RocksDB, I don't recall it precisely

I don't think this is only my issue, but it should be a common issue as long as user has set ttl. 

"The modification will make every compaction go through CompactionFilter, it will hurt performance a lot"

> Do we have performance testing for such cases? Personally I doubt this is the fact according to my experience in the production environment. 

And also please trust me, changing periodic_compaction_seconds won't help at all. I tested this a lot of times.

Reply:
> Do we have performance testing for such cases? Personally I doubt this is the fact according to my experience in the production environment.

If you do so some profiling, you will find that custom compaction filter is way more slow that default one. Because for each key, it will check the schema is valid or not, parse the value and check ttl. Validation of schema is the bottleneck.

Reply:
> changing periodic_compaction_seconds won't help at all.

How do you set it? And why that rocksdb didn't pick the outdate sst file

Reply:
> > changing periodic_compaction_seconds won't help at all.
> 
> How do you set it? And why rocksdb didn't pick the outdate sst file at bottom level, the score is not big enough or what?
> 
> IIRC, if your file has been `marked_for_compaction`, it definitely will be picked. The question is that have you marked them

Yes, I changed the periodic_compaction_seconds and DO noticed that the bottom level files had all gone through compaction. But it won't help at all because it is just doing default compaction, without going through custom compaction filter. The default compaction doesn't have the logic to validate if the data is expired, so as you can see,  the data will still be there.

Reply:
If the bottom sst are picked, I agree to remove `custom_filter_interval_secs`, but we need to address the performance issue as well.

Here is the way to leverage performance:
1. StorageCompactionFilter have constructor of SchemaManager and IndexManager
2. Whenever we check a key/value could pass the filter, we check whether the schema are valid. So for each key, we will check it like codes below
```
    auto tagId = NebulaKeyUtils::getTagId(vIdLen_, key);
    auto schema = schemaMan_->getTagSchema(spaceId, tagId);
    if (!schema) {
      VLOG(3) << "Space " << spaceId << ", Tag " << tagId << " invalid";
      return false;
    }
```
edges are similar.
3. What we need to do is that, when we construct StorageCompactionFilter, get every schema once and for all (by calling `getAllVerTagSchema` and `getAllVerEdgeSchema` in SchemaManager). When we valid schema later, we only need to check it locally instead of check it in SchemaManager.

Reply:
See the latest comment, there is a simpler way to address the issue.